### ISSUE TITLE
Explore migrating from pymongo to motor as an async python libary for mongodb

### ISSUE THREAD

**Issue by mkshah605** [microbiomedata/nmdc-runtime#787]
Given that there is now "official" guidance from MongoDB Inc. on using their official async python library for mongodb, i.e. [motor](https://pypi.org/project/motor/), with FastAPI, explore pros/cons of migrating from using the older, more-stable-but-not-async [pymongo](https://pypi.org/project/pymongo/) library in order to take full performance advantage of FastAPI's async support.

The implication of this would be that a single-container deployment of the runtime would: 
- support a higher rate of requests (avoid having to run multiple containers)
- in turn, potentially help avoid having to consider race conditions wrt multiple runtime containers accessing the same backing mongodb instance

**Goal**: take some time to compare both approaches to better inform a path forward

**Comment by eecavanna** [comment]
Hello, @mkshah605 — welcome to the project!

@mkshah605 and @dwinston — While working on migrating `nmdc-runtime` from `pip-tools` to Poetry, I came across both `motor` and `pymongo` in the dependency tree. Based on the digging I did at the time, I (to my surprise) came away thinking that the Runtime was actually already using `motor` and it was only some miscellaneous experiments that were using `pymongo`.

https://github.com/microbiomedata/nmdc-runtime/blob/d7da63fdfaf9262ad1da4d0d9c4301b8d119b534/nmdc_runtime/api/db/mongo.py#L15

So, today, I was surprised when I noticed this ticket in the repo.

**Comment by eecavanna** [comment]
Oops... correction: it is using both (and I think it's using `pymongo` for most things—as opposed to the opposite, which I suggested in my previous message). 

https://github.com/microbiomedata/nmdc-runtime/blob/d7da63fdfaf9262ad1da4d0d9c4301b8d119b534/nmdc_runtime/api/db/mongo.py#L14-L15

https://github.com/microbiomedata/nmdc-runtime/blob/d7da63fdfaf9262ad1da4d0d9c4301b8d119b534/nmdc_runtime/api/db/mongo.py#L27-L28

Maybe someone was experimenting with `motor` and left their experimentation in the code base.

**Comment by dwinston** [comment]
@eecavanna yes, @elais was leading the experimentation with `motor` before he was eaten by JAWS... 🦈

**Comment by mkshah605** [comment]
Thank you @eecavanna !

Yes, after looking into this, it seems like `motor` is the way to go. It is functionally equivalent to `pymongo`, and is designed for asynchronous frameworks like asyncio. Especially since we already have some code that uses motor, it might be worth doing a full refactor for consistency.

I will create a ticket for this and link it here.

**Comment by dwinston** [comment]
[`pymongo` now has a stable (no longer in beta) asynchronous API](https://pymongo.readthedocs.io/en/stable/changelog.html#changes-in-version-4-13-0-2025-05-14). See the [pymongo.asynchronous](https://pymongo.readthedocs.io/en/stable/api/pymongo/asynchronous/index.html#module-pymongo.asynchronous) docs or the [migration guide](https://www.mongodb.com/docs/languages/python/pymongo-driver/current/reference/migration/) for more information.

**Comment by shreddd** [comment]
@dwinston - could you scope what this would take? If it is reasonable effort please open a new issue, otherwise we can close and revisit in the future if needed.

**Comment by eecavanna** [comment]
I'm working on this in PR #1145.

The set of functions requiring updates continues to balloon, as I think we all expected, but I just now ran into one that I am not sure I will be able to update: a function named `persist_content_and_get_drs_object` calls `GridFS` (imported from `gridfs`) and passes it a `Database`. I am not sure `GridFS` can accept an `AsyncDatabase` in its place. Anyway, I wanted to note this before moving on. Right now, I'm trying to update only the `/nmdcschema/{collection_name}` endpoint, although doing so it involving updating lots of functions and other endpoints because the "workhorse" function the endpoint uses is used by other functions/endpoints also. Work in progress.
